### DISCUSSION TITLE
How do I return a ReadableStream in an API route?

### DISCUSSION THREAD

**Discussion by vedantroy** [vercel/next.js#50614] [Help]
### Summary

I want to return a file in an API route. However, I don't know how to do that. Well.... this works:
```
  const buf = await fs.promises.readFile("example.txt")
  return new NextResponse(ReadableStream.)
```

but, I want to return the file in a streaming fashion. This means, I need to use something like the `ReadableStream` API. Not sure how to do this.

**Comment by icyJoseph** [comment]
You mean like this? https://nextjs.org/docs/app/building-your-application/routing/router-handlers#streaming

**Reply by vedantroy** [reply]
Oh this seems like the thing i'm looking for, I'm surprised there is not a builtin nodejs stream to nextjs readable stream though

**Reply by nmitic** [reply]
I do not know how the link above helped you? I have a stream which I want to return and later consume on the client, link doesn't explain how to do that, instead it uses a bunch of helper func from openai react package.

**Reply by lukeocodes** [reply]
I think there needs to be a better way to do this, especially if you don't intend on streaming text back.

**Comment by OliverDudgeon** [comment]
Is there any way to do this with API routes in the **pages** router? It wants a `NodeJS.WritableStream` but I just have a `ReadableStream` transformed from a `fetch(...)` call?

**Comment by stalingino** [comment]
```js
export async function GET(req) {
	const filePath = path.join(process.cwd(), 'public', req.nextUrl.searchParams.get('fileName') + '.html');
	const res = new NextResponse(fs.createReadStream(filePath, { encoding: 'utf8' }));
	res.headers.set('Content-Type', 'text/html');
	return res;
}
```
This is working very well.

**Reply by jasxir** [reply]
I got this.
```
Argument of type 'ReadStream' is not assignable to parameter of type 'BodyInit | null | undefined'.
  Type 'ReadStream' is missing the following properties from type 'URLSearchParams': size, append, delete, get, and 8 more.ts(2345)
```

**Reply by sommeeeer** [reply]
@jasxir 
```ts
// for pages router
res.setHeader("Content-Type", "text/html; charset=utf-8");
res.setHeader("Connection", "keep-alive");
res.setHeader("Cache-Control", "no-cache, no-transform");
return Readable.fromWeb(stream).pipe(res);

// for app router
return new Response(stream, {
    headers: {
      "Content-Type": "text/html; charset=utf-8",
      Connection: "keep-alive",
      "Cache-Control": "no-cache, no-transform",
    },
  });
```

**Reply by sahibjotsaggu** [reply]
@sommeeeer for some reason, the `Connection: "keep-alive",` header doesn't show up in production, but works in dev. Any ideas?

**Reply by sommeeeer** [reply]
> @sommeeeer for some reason, the `Connection: "keep-alive",` header doesn't show up in production, but works in dev. Any ideas?

If you are using Vercel, or AWS it is remapped to a different header called `x-amzn-remapped-connection`. My guess is that Vercel completely leaves it out. Its restricted in many serverless environments. Is it there if you do `next start` or `next .next/standalone/server.js`. Then you can be certain the `Connection` header is stripped out. Its more common than not.

**Reply by sahibjotsaggu** [reply]
Hmm, I've never seen that header. Confirming I don't see it on my vercel production deployment. Locally, I see the same result with `next dev` and `next start`. I've been looking into hosting my long lived API elsewhere to get around this. Unfortunate, but I feel like I'm missing some config somewhere.

**Reply by sommeeeer** [reply]
> I've been looking into hosting my long lived API elsewhere to get around this. Unfortunate, but I feel like I'm missing some config somewhere.

To be honest, if your just streaming responses, I dont think most modern web browsers would care about that header. It should just work anyways. Else those deployment platforms would not strip it out, I beieve.

**Reply by TechQuery** [reply]
@sommeeeer @sahibjotsaggu `keep-alive` may be not necessary?
https://www.ericburel.tech/blog/nextjs-stream-files#how-to-stream-big-files-the-old-way-with-api-routes-nextjs-9

**Reply by sahibjotsaggu** [reply]
@TechQuery yes, after looking into this more, I realized I didn't need many of the headers to do the streaming. In fact, I don't have any special headers for my AI streaming.

**Reply by sommeeeer** [reply]
@TechQuery yeah as i said, its not really mandatory. you can most likely skip it. what is is important tho is that you close the stream, which is done automatically by Next if you return it, or you would have to call `res.end()` your self

**Reply by TechQuery** [reply]
> what is is important tho is that you close the stream, which is done automatically by Next if you return it, or you would have to call res.end() your self

@sommeeeer `return` or `res.end()` may be not necessary too?
https://nextjs.org/docs/pages/building-your-application/routing/api-routes
